### PR TITLE
feat(sourcemaps): Add endpoint for artifact bundles files

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR implements a new endpoint for fetching the files inside a specific article bundle.

### CODE REVIEW

File: src/sentry/api/urls.py [hunk 1]
```diff
@@ -1963,6 +1964,11 @@
         ProjectReleaseStatsEndpoint.as_view(),
         name="sentry-api-0-project-release-stats",
     ),
+    url(
+        r"^(?P<organization_slug>[^\/]+)/(?P<project_slug>[^\/]+)/artifact-bundle/(?P<bundle_id>[^/]+)/files/$",
```
Comment (kamilogorek) [inline src/sentry/api/urls.py:1968]: I'd double check that this is the url we want to use. All urls in this kind of shape are using plurals if they allow for specifying an ID. eg.

```
/org/projects/releases/<release-id>/files/
/files/dsyms/
/files/source-maps/
/files/artifact-bundles/
/files/difs/
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+class ArtifactBundleSource:
+    def __init__(self, files: dict):
+        self._files = files
+
+    @cached_property
+    def sorted_and_filtered_files(self) -> List[Tuple[str, dict]]:
+        files = [(file_path, info) for file_path, info in self._files.items()]
+        files.sort(key=lambda item: item[0])
+
+        return files
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:26]:
```suggestion
        return sorted(list(_files.items()))
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+@region_silo_endpoint
+class ProjectArtifactBundleFilesEndpoint(ProjectEndpoint):
+    permission_classes = (ProjectReleasePermission,)
+    rate_limits = RateLimitConfig(
+        group="CLI", limit_overrides={"GET": SENTRY_RATELIMITER_GROUP_DEFAULTS["default"]}
+    )
+
+    def get(self, request: Request, project, bundle_id) -> Response:
+        """
+        List a Project Artifact Bundle's Files
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:44]:
```suggestion
        List files for a given project artifact bundle
```

Maybe more readable this way?

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        ``````````````````````````````
+
+        Retrieve a list of artifact bundle files for a given artifact bundle.
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:47]:
```suggestion
        Retrieve a list of files for a given artifact bundle.
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        :pparam string organization_slug: the slug of the organization the
+                                          release belongs to.
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:50]:
```suggestion
                                          artifact bundle belongs to.
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        :pparam string project_slug: the slug of the project to list the
+                                     release files of.
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:52]:
```suggestion
        :pparam string project_slug: the slug of the project the
                                     artifact bundle belongs to.
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        :pparam string bundle_id: bundle_id of the artifact bundle to read contents from.
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:53]:
```suggestion
        :pparam string bundle_id: bundle_id of the artifact bundle to list files from.
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            return Response(
+                {"error": f"The artifact bundle with {bundle_id} is not bound to this project"},
+                status=400,
+            )
+
+        try:
+            # We open the archive to fetch the number of files.
+            archive = ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=False)
+        except Exception:
+            return Response({"error": "The artifact bundle archive can't be opened"})
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:79]:
```suggestion
            return Response({"error": "The archive of artifact bundle {bundle_id} can't be opened"})
```

File: src/sentry/api/endpoints/project_artifact_bundle_files.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                "size": file_info.file_size if file_info is not None else None,
+            }
+
+        def serialize_results(r):
+            serialized_results = serialize(
+                [expose_artifact_bundle_file(file_path, info) for file_path, info in r],
+                request.user,
+            )
+            # We must close the archive once all the results have been fetched, otherwise we will get an error.
+            archive.close()
```
Comment (kamilogorek) [inline src/sentry/api/endpoints/project_artifact_bundle_files.py:101]: Is it possible that this flow (or `expose_artifact_bundle_file` itself) will somehow fail along the way and archive will be kept opened? Can we do something like `finally` in python? Is it also possible that `paginate` will fail and we should handle this case and close archive then as well?

**Review by kamilogorek — CHANGES_REQUESTED** [review]

File: src/sentry/api/endpoints/project_artifact_bundle_files.py (hunk 9 above)
Comment (iambriccardo) [reply src/sentry/api/endpoints/project_artifact_bundle_files.py:101]: I hate doing this `.close` here and was wondering the same thing. I might have to see what we can do in cases of failure. I might wrap the pagination with an error handling in place, or actually, close the archive before returning the resultset.

**Review by kamilogorek — APPROVED** [review]

**Pull request merged**